File: packages/core/deepsolo/models/strategy.py
"""策略数据模型，与前端 Strategy 接口对齐"""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class StrategyData:
    """策略表现数据，对应前端 Strategy 接口"""
    id: str
    name: str
    category: str             # "hot" | "normal" | "emerged"
    description: str
    return_pct: float         # 对应前端 returnPct
    max_drawdown_pct: float   # 对应前端 maxDrawdownPct
    total_trades: int
    win_rate: float
    avg_return_pct: float
    capital: float
    state: str                # "idle" | "backtesting" | "evolving" | "competing" | "discussing" | "profitable" | "retired"
    parents: list[str] = field(default_factory=list)
    relation: str | None = None

    def to_frontend_dict(self) -> dict:
        """生成前端 Strategy 接口格式的 dict"""
        return {
            "id": self.id,
            "name": self.name,
            "category": self.category,
            "description": self.description,
            "returnPct": self.return_pct,
            "maxDrawdownPct": self.max_drawdown_pct,
            "totalTrades": self.total_trades,
            "winRate": self.win_rate,
            "avgReturnPct": self.avg_return_pct,
            "capital": self.capital,
            "state": self.state,
            "parents": self.parents,
            "relation": self.relation,
        }

    @classmethod
    def from_frontend_dict(cls, d: dict) -> StrategyData:
        """从前端 Strategy 格式反序列化"""
        return cls(
            id=d["id"],
            name=d["name"],
            category=d["category"],
            description=d["description"],
            return_pct=d["returnPct"],
            max_drawdown_pct=d["maxDrawdownPct"],
            total_trades=d["totalTrades"],
            win_rate=d["winRate"],
            avg_return_pct=d["avgReturnPct"],
            capital=d["capital"],
            state=d.get("state", "idle"),
            parents=d.get("parents", []),
            relation=d.get("relation"),
        )

File: packages/visual/src/types/strategy.ts
/** Agent 状态枚举 */
export enum AgentState {
  Idle = 'idle',
  Backtesting = 'backtesting',
  Evolving = 'evolving',
  Competing = 'competing',
  Discussing = 'discussing',
  Profitable = 'profitable',
  Retired = 'retired',
}

/** 策略类别 */
export type StrategyCategory = 'hot' | 'normal' | 'emerged';

/** 策略数据 */
export interface Strategy {
  id: string;
  name: string;
  category: StrategyCategory;
  description: string;
  returnPct: number;
  maxDrawdownPct: number;
  totalTrades: number;
  winRate: number;
  avgReturnPct: number;
  capital: number;
  state: AgentState;
  parents?: string[];
  relation?: string;
  buildingId?: string;
  placement?: 'world' | 'indoor-only';
  role?: string;
  sourceWorkspace?: string;
  mode?: string;
}

/** 策略模板 (来自 template.json) */
export interface StrategyTemplate {
  path: string;
  name: string;
  description: string;
  return_pct: number;
}

/** 账户汇总 (来自 v*_account.json) */
export interface AccountSummary {
  start_date: string;
  end_date: string;
  final_capital: number;
  total_return_pct: number;
  max_drawdown_pct: number;
  total_trades: number;
  win_trades: number;
  lose_trades: number;
  win_rate: number;
  avg_return_pct: number;
}

/** 账户数据 */
export interface AccountData {
  config: { initial_capital: number; max_holdings: number };
  summary: AccountSummary;
}

/** 事件日志条目 */
export interface EventEntry {
  time: string;
  agentName: string;
  text: string;
}

/** 状态区域配置 */
export interface StateRegion {
  x: number;
  y: number;
  radius: number;
}

/** 状态标签映射 */
export const STATE_LABELS: Record<AgentState, string> = {
  [AgentState.Idle]: '待命',
  [AgentState.Backtesting]: '回测',
  [AgentState.Evolving]: '进化',
  [AgentState.Competing]: '竞争',
  [AgentState.Discussing]: '讨论',
  [AgentState.Profitable]: '盈利',
  [AgentState.Retired]: '淘汰',
};

/** 状态转换规则 */
export const STATE_TRANSITIONS: Record<AgentState, AgentState[]> = {
  [AgentState.Idle]: [AgentState.Backtesting, AgentState.Discussing],
  [AgentState.Backtesting]: [AgentState.Evolving, AgentState.Competing, AgentState.Idle],
  [AgentState.Evolving]: [AgentState.Backtesting, AgentState.Competing],
  [AgentState.Competing]: [AgentState.Profitable, AgentState.Discussing, AgentState.Idle],
  [AgentState.Discussing]: [AgentState.Competing, AgentState.Idle],
  [AgentState.Profitable]: [AgentState.Competing, AgentState.Retired],
  [AgentState.Retired]: [],
};
